### ISSUE TITLE
Feature: `pnpm run` preserve current working directory

### ISSUE THREAD

**Issue by shellscape** [pnpm/pnpm#5464]
- [x] I'd be willing to implement this feature ([contributing guide](https://github.com/pnpm/pnpm/blob/main/CONTRIBUTING.md))

**Describe the user story**

When running a workspace-root package script from a workspace package's script, I would like to instruct the rest of the, or at least next, `pnpm run` in the call chain to use the original current working directory.

**Describe the solution you'd like**

`--preserve-cwd` could work as a name

**Describe the drawbacks of your solution**

This may get hairy when chaining multiple `pnpm run` calls, but if we mandated that only the immediate call with `--preserve-cwd` retained the CWD, I think we could get around trying to be too intelligent about it. 

**Describe alternatives you've considered**

I've tried using `$npm_package_name` and setting environment variables like `SERVICE_PATH=$(basename $(dirname $PWD))` but it's a pill to manage, and ends up with boilerplate all over the place.

**Comment by shellscape** [comment]
@zkochan I've started looking into this, but I get lost around `runLifecycleHook` which seems to be calling itself recursively via the `lifecycle` default import. I _think_ I need to start at `parseCliArgs` to add the new option, and then it gets tricked down? And I might need to affect `opts.rootModulesDir`? Not sure where to go to get this done, as the commands stuff is very twisty and turny. Any help in implementation would be greatly appreciated.

**Comment by shellscape** [comment]
I described why in the alternatives section of the template. without the ability to preserve the current working directory when executing workspace scripts from a package directory, you have to make a mess out of environment variables in manually change directories. it's an incredible amount of boilerplate

**Comment by shellscape** [comment]
End goal is to make root package.json scripts use the same working directory as the workspace package.json script that is invoking the root script. I'll paste some json examples in a bit.

**Comment by zkochan** [comment]
ok, I guess the simplest solution would be just to merge the scripts from the root to the scripts of every workspace package.json. Then you would just run the scripts as if they were the scripts of the workspace.

**Comment by shellscape** [comment]
yeah that would be against what we're trying to accomplish, and that's less code, not more. DRY and all. that's more complex than using env vars to set the CWD due to duplication

**Comment by zkochan** [comment]
I mean pnpm would do that in memory.

**Comment by shellscape** [comment]
Ah I see. Seems like it would be more complicated when running recursive scripts, because we intentionally ignore the root now. So there would have to be concessions made for that.

Here's the primary use case we're looking at, for posterity:

```js
  // /packages/batman/package.json
  "scripts": {
    "build": "tsc --project tsconfig.json",
    "postbuild": "pnpm -w package:postbuild",
    "prebuild": "pnpm -w package:prebuild" 
  },
```

```js
  // /package.json
  "scripts": {
    "package:postbuild": "rm tsconfig.json",
    "package:prebuild": "ln -s ../../shared/tsconfig.packages.json tsconfig.json"
  },
```

So we can see that the goal is creating an intermediate config file that a process will use, and then discarding it once the process is done. One might say, why not do this in every package:

```js
  // /packages/batman/package.json
  "scripts": {
    "build": "tsc --project tsconfig.json",
    "postbuild": "rm tsconfig.json",
    "prebuild": "ln -s ../../shared/tsconfig.packages.json tsconfig.json" 
  },
```

We have ~30 packages, and we've changed the location and filename of the target config file a few times now, which was not fun updating all packages. There are tools like `mrm` which could facilitate that for us, yes. Then there's the matter of DRY, which we try to stick to as a philosophy when possible. So that leads to a solution like this:

```js
  // /packages/batman/package.json
  "scripts": {
    "build": "tsc --project tsconfig.json",
    "postbuild": "pnpm -w package:postbuild",
    "prebuild": "pnpm -w package:prebuild" 
  },
```

```js
  // /package.json
  "scripts": {
    "package:postbuild": "rm tsconfig.json",
    "package:prebuild": "ln -s ../../shared/tsconfig.packages.json tsconfig.json"
  },
```

But `build` runs in `/packages/batman` and `package:prebuild` runs in `/`. So that won't work. The next step is:

```js
  // /packages/batman/package.json
  "scripts": {
    "build": "tsc --project tsconfig.json",
    "postbuild": "PKGDIR=$PWD pnpm -w package:postbuild",
    "prebuild": "PKGDIR=$PWD pnpm -w package:prebuild" 
  },
```

```js
  // /package.json
  "scripts": {
    "package:postbuild": "cd $PKGDIR && rm tsconfig.json",
    "package:prebuild": "cd $PKGDIR && ln -s ../../shared/tsconfig.packages.json tsconfig.json"
  },
```

That achieves the step-separation we're after and lets the root script run in the package script CWD, but she's ugly. This is where the `--preserve-cwd` idea was born. I'm going to keep trying variations to reduce that code footprint in scripts, but we've run into the same kind of scenario in a few other spots that are more difficult to form elegant solutions and the option would remove a lot of code on our end.

Whether or not it's worth added complexity in pnpm I leave to you :smile:

**Comment by zkochan** [comment]
so you would just do this:

```json
  // /packages/batman/package.json
  "scripts": {
    "build": "tsc --project tsconfig.json",
    "postbuild": "pnpm package:postbuild",
    "prebuild": "pnpm package: prebuild" 
  },
```

**Comment by shellscape** [comment]
yeah that could work
